### PR TITLE
Filter with cached count in prewhere

### LINKED ISSUE #40956: PREWHERE filters may be ignored when row policy used
**Describe the unexpected behaviour**
Using a WITH statement when row policy is applied to the table leads to unexpected result where rows are not getting filtered how they should be.

**How to reproduce**
* Which ClickHouse server version to use
`ClickHouse server version 22.8.2 revision 54460`

* Queries to run that lead to unexpected result
```
): CREATE TABLE table0
(
    `x` Int16,
    `y` String
)
ENGINE = MergeTree
ORDER BY x
SETTINGS index_granularity = 8192

): INSERT INTO table0(y, x) VALUES ('A',1), ('B',2), ('C',3)

): CREATE ROW POLICY pol0 ON default.table0 FOR SELECT USING x >= 0 TO default

): WITH Test AS
    (
        SELECT
            multiIf(y = 'Aaaaaaaaaaaa', 'Bbbbbbbbbbbb', y) AS y,
            x
        FROM table0
        WHERE y IN ('A', 'C')
    )
SELECT DISTINCT (y, x)
FROM Test

Query id: 0b494596-392a-47cf-8587-239561219844

┌─tuple(y, x)─┐
│ ('A',1)     │
│ ('B',2)     │
│ ('C',3)     │
└─────────────┘
```

**Expected behavior**
Only 'A' and 'C' entries should appear, 'B' should be excluded by the `WHERE y IN ('A', 'C')` clause.

### PR DESCRIPTION
**Pull request by davenger**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix PREWHERE with row-level filters (when row filter is always true/false)

In case of row-level filters optimized out, i.e. converted to always true/false, it is possible for MergeTreeRangeReader to reuse incorrect statistics for the filter (countBytesInResultFilter()), and because of this it simply does not apply other filters, since it assume that this filter does not need to filter anything.

This is a reworked fix by @azat #41940
Fixes: #40956

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by azat** [comment]
@davenger are you going to continue to this?

File: src/Storages/MergeTree/MergeTreeRangeReader.h [hunk 1]
```diff
@@ -230,17 +260,13 @@ class MergeTreeRangeReader
...
-        const ColumnUInt8 * filter = nullptr;
-        const ColumnUInt8 * filter_original = nullptr;
+        FilterWithCachedCount filter;
+        FilterWithCachedCount filter_original;
 
         void collapseZeroTails(const IColumn::Filter & filter, IColumn::Filter & new_filter);
         size_t countZeroTails(const IColumn::Filter & filter, NumRows & zero_tails, bool can_read_incomplete_granules) const;
         static size_t numZerosInTail(const UInt8 * begin, const UInt8 * end);
 
-        std::map<const IColumn::Filter *, size_t> filter_bytes_map;
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/Storages/MergeTree/MergeTreeRangeReader.h:242]: IIRC, the map is only for simplicity. This `filter_bytes_map` is even older than we have row-level filters)

**Comment by filimonov** [comment]
@davenger any progress here?

**Comment by davenger** [comment]
This fix was included in https://github.com/ClickHouse/ClickHouse/pull/42126

**Pull request closed without merging**